File: packages/terse-cli/src/deploy.ts
import chalk from "chalk"
import { zipSync } from "fflate"
import fs from "node:fs"
import path from "node:path"
import ora from "ora"
import { ApiRoutes, sdkDeployRequestBodySchema } from "terse-types"
import type { SdkDeployResponseBody } from "terse-types"

import { fetchWithAuth, readApiKeyOrBail, readEnvVar } from "./api.js"
import { loadJobRegistry } from "./loadJob.js"
import type { LanguageProvider } from "./providers/LanguageProvider.js"
import { resolveProvider } from "./providers/resolveProvider.js"

export async function deploy(provider: LanguageProvider = resolveProvider(), entryFile?: string) {
    const apiKey = readApiKeyOrBail({
        title: "Error: No TERSE_API_KEY found in .env",
        detail: "Run `terse init` to set up your project, or add TERSE_API_KEY to your .env file."
    })

    const registry = await loadJobRegistry(provider, entryFile)
    const jobs = [...registry.values()]

    // If TERSE_JOB_URL is set in .env, deploy in URL mode (user infrastructure).
    // All jobs get this URL and no source code is zipped or uploaded.
    const jobUrl = readEnvVar("TERSE_JOB_URL")
    const isUrlMode = !!jobUrl

    let sourceZipBase64: string | undefined
    let fileCount = 0
    let zipSizeBytes = 0

    if (!isUrlMode) {
        const zipPayload = buildZipPayload(provider)
        sourceZipBase64 = zipPayload.sourceZipBase64
        fileCount = zipPayload.fileCount
        zipSizeBytes = zipPayload.zipSizeBytes
    }

    const spinner = ora(`Deploying ${jobs.length} job${jobs.length === 1 ? "" : "s"}...`).start()

    try {
        const body = sdkDeployRequestBodySchema.parse({
            jobs: jobs.map(job => ({
                jobName: job.name,
                triggers: job.triggers,
                outputs: job.skills ?? [],
                toolApprovals: job.toolApprovals ?? []
            })),
            jobUrl: isUrlMode ? jobUrl : undefined,
            sourceZipBase64
        })

        const result = await fetchWithAuth<SdkDeployResponseBody>(ApiRoutes.SDK.DEPLOY, apiKey, body, "POST")

        if (!result.success) {
            spinner.fail(chalk.red(`Deploy failed: ${result.error}`))
            if (result.details) {
                console.error(chalk.dim(`  ${result.details}`))
            }
            process.exit(1)
        }

        spinner.succeed(chalk.green(`Deployed ${result.results.length} job${result.results.length === 1 ? "" : "s"}`))

        for (const r of result.results) {
            const verb = r.isUpdate ? "Updated" : "Created"
            console.log(chalk.dim(`  ${verb} "${r.jobName}" (${r.automationId})`))
            if (r.triggers) {
                for (const t of r.triggers) {
                    if (t.metadata?.webhookUrl) {
                        console.log(chalk.cyan(`    Webhook URL: ${t.metadata.webhookUrl}`))
                    }
                }
            }
        }

        if (isUrlMode) {
            console.log(chalk.dim(`  Mode: user infrastructure`))
            console.log(chalk.dim(`  Job URL: ${jobUrl}`))
        } else {
            console.log(chalk.dim(`  Files: ${fileCount}`))
            console.log(chalk.dim(`  Zip size: ${(zipSizeBytes / 1024).toFixed(1)} KB`))
        }

        if (result.removed.length > 0) {
            console.log(chalk.yellow(`\nRemoved ${result.removed.length} stale job${result.removed.length === 1 ? "" : "s"} no longer in project:`))
            for (const r of result.removed) {
                console.log(chalk.dim(`  - ${r.name} (${r.id})`))
            }
        }
    } catch (error) {
        spinner.fail(chalk.red(`Deploy failed: ${(error as Error).message}`))
        process.exit(1)
    }
}

function collectFiles(dir: string, baseDir: string, provider: LanguageProvider): Record<string, Uint8Array> {
    const entries: Record<string, Uint8Array> = {}

    for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
        if (provider.deployExclusions.dirs.has(entry.name)) continue
        if (provider.deployExclusions.files.has(entry.name)) continue

        const fullPath = path.join(dir, entry.name)
        const relativePath = path.relative(baseDir, fullPath)

        if (entry.isDirectory()) {
            Object.assign(entries, collectFiles(fullPath, baseDir, provider))
        } else if (entry.isFile()) {
            entries[relativePath] = new Uint8Array(fs.readFileSync(fullPath))
        }
    }

    return entries
}

function buildZipPayload(provider: LanguageProvider): { sourceZipBase64: string; fileCount: number; zipSizeBytes: number } {
    const cwd = process.cwd()
    const files = collectFiles(cwd, cwd, provider)
    const fileCount = Object.keys(files).length

    if (fileCount === 0) {
        console.error(chalk.red("No files found to deploy"))
        process.exit(1)
    }

    const zipData = zipSync(files, { level: 6 })
    return {
        sourceZipBase64: Buffer.from(zipData).toString("base64"),
        fileCount,
        zipSizeBytes: zipData.length
    }
}

File: packages/terse-python-cli/src/terse_cli/commands/deploy.py
"""`terse deploy` command."""

from __future__ import annotations

import click
from pydantic import ValidationError
from terse_sdk import (
    RegisteredJob,
    SdkDeployJob,
    SdkDeployRequestBody,
    SdkDeployResponseBody,
    SdkDeployTrigger,
    TriggerConfig,
)

from .._http import ApiRequestError, AuthenticationError, request_json
from .._loader import NoJobsFoundError, ProjectImportError, load_job_registry
from .._package import PackagingError, build_deploy_archive
from .._project import ProjectRootError, assert_project_root, read_api_key
from .._ui import console


@click.command("deploy", help="Deploy all jobs to Terse (syncs with server; removed jobs are deleted).")
def deploy_command() -> None:
    try:
        project_dir = assert_project_root()
        api_key = read_api_key(project_dir)
        if not api_key:
            raise click.ClickException(
                "No TERSE_API_KEY found in .env.\n"
                "Run `terse init` to set up your project, or add TERSE_API_KEY to your .env file."
            )

        _, registry = load_job_registry(project_dir)
        archive = build_deploy_archive(project_dir)
        jobs = list(registry.values())

        with console.status(f"Deploying {len(jobs)} job{'s' if len(jobs) != 1 else ''}..."):
            payload = request_json(
                "/sdk/deploy",
                api_key,
                method="POST",
                params=SdkDeployRequestBody(
                    jobs=[_serialize_job(job) for job in jobs],
                    sourceZipBase64=archive.source_zip_base64,
                ).model_dump(exclude_none=True),
            )

        response = SdkDeployResponseBody.model_validate(payload)
    except (ProjectRootError, ProjectImportError, NoJobsFoundError, PackagingError) as exc:
        raise click.ClickException(str(exc)) from exc
    except AuthenticationError as exc:
        raise click.ClickException(
            "Authentication failed: your TERSE_API_KEY was rejected.\nUpdate TERSE_API_KEY in .env and try again."
        ) from exc
    except ApiRequestError as exc:
        raise click.ClickException(str(exc)) from exc
    except ValidationError as exc:
        raise click.ClickException(f"Received an invalid deploy response.\n\n{exc}") from exc

    if not response.success:
        details = f"\n\n{response.details}" if response.details else ""
        raise click.ClickException(f"Deploy failed: {response.error or 'unknown error'}{details}")

    console.print(f"[green]Deployed {len(response.results)} job{'s' if len(response.results) != 1 else ''}[/green]")
    for result in response.results:
        verb = "Updated" if result.isUpdate else "Created"
        console.print(f'  {verb} "{result.jobName}" ({result.automationId})')

    console.print(f"  Files: {archive.file_count}")
    console.print(f"  Zip size: {archive.zip_size_bytes / 1024:.1f} KB")

    if response.removed:
        console.print("")
        console.print(
            f"[yellow]Removed {len(response.removed)} stale job{'s' if len(response.removed) != 1 else ''}[/yellow]"
        )
        for removed in response.removed:
            console.print(f"  {removed.name} ({removed.id})")


def _serialize_job(job: RegisteredJob) -> SdkDeployJob:
    return SdkDeployJob.model_validate(
        {
            "jobName": job.name,
            "triggers": [_serialize_trigger(trigger).model_dump(exclude_none=True) for trigger in job.triggers],
            "webhookURL": job.webhook_url,
        }
    )


def _serialize_trigger(trigger: TriggerConfig) -> SdkDeployTrigger:
    return SdkDeployTrigger.model_validate(
        {
            "configType": trigger.config_type or "",
            "integrationType": trigger.integration_type,
            "integrationId": trigger.integration_id,
            "config": dict(trigger.config),
        }
    )
